internal/worker/concurrency.go

```diff
@@ -74,7 +77,7 @@ func (p *permit) AcquireChan(ctx context.Context, wg *sync.WaitGroup) <-chan str
 		}
 		select { // try to send to channel, but don't block if listener is gone
 		case ch <- struct{}{}:
-		default:
+		case <-time.After(10 * time.Millisecond): // wait time is needed to avoid race condition of channel sending
```

Comment:
found a race condition of channel sending. Adding a wait time would be more reliable

Reply:
hmm :\  unfortunately this means that if processing is delayed by 10ms it will block the chan-reader forever.  that's not too unlikely with big CPU spikes, and definitely not impossible.

tbh I think that might rule this impl out entirely.  though I think it's possible to build a `AcquireChan(...) (<-chan struct{}, cancel func())` that doesn't have this issue, and that might be worth doing.

Reply:
if we expect chan-reader to always consume from the returned `ch` unless ctx is canceled then we can replace this goroutine implementation with 

```
defer wg.Done()
if err := p.sem.Acquire(ctx, 1); err != nil { 
   return // assuming Acquire only returns err if ctx.Done
}
select {
case ch <- struct{}{}:
case <-ctx.Done():
  p.sem.Release(1)
}
```

Reply:
fixed through Taylan's suggestion. It should be safe now.